ludwig/models/trainer.py

```diff
@@ -331,6 +335,230 @@ def write_step_summary(
+                        for i_feat in model.input_features.values()
+                    }
+                    targets = {
+                        o_feat.feature_name: batch[o_feat.proc_column]
+                        for o_feat in model.output_features.values()
+                    }
+
+                    loss, _ = model.train_step(
+                        self.optimizer,
+                        inputs,
+                        targets,
+                        self.regularization_lambda
+                    )
+                    # compute smoothed loss
+                    avg_loss = beta * avg_loss + (1-beta) * loss
+                    smoothed_loss = avg_loss / (1 - beta**(step_count + 1))
+
+                    # store learning rate and loss
+                    learning_rates.append(current_learning_rate)
+                    losses.append(smoothed_loss)
+
+                    # check whether loss is diverging
+                    if step_count > 0 and smoothed_loss > early_stop_threshold * best_loss:
+                        diverging = True
+                        break
+                    else:
+                        if smoothed_loss < best_loss or step_count == 0:
+                            best_loss = smoothed_loss
+
+                    # compute new learning rate
+                    if mode == "exponential":
+                        current_learning_rate = exponential_scheduler(
+                            current_learning_rate, step_count)
+                    else:
+                        current_learning_rate = linear_scheduler(
+                            current_learning_rate, step_count)
+
+                    self.optimizer.set_learning_rate(current_learning_rate)
+                    step_count += 1
+
+                epoch += 1
+
+        optimal_lr = get_optimal_lr(losses, learning_rates)
+        if optimal_lr:
+            self.learning_rate = optimal_lr
+        return self.learning_rate
+
+    def tune_batch_size(
+        self,
+        config,
+        training_set,
+        random_seed: int = default_random_seed,
+        max_trials: int = 10,
+        halving_limit: int = 3
+    ):
+        from ludwig.api import LudwigModel
+
+        def _is_valid_batch_size(batch_size):
+            return batch_size < len(training_set)
+
+        # TODO (ASN) : Circle back on how we want to set default placeholder value
+        # Currently, since self.batch_size is originally set to auto, we provide a
+        # placeholder starting value (namely, 128)
+        self.batch_size = 128
+        skip_save_model = self.skip_save_model
+        skip_save_progress = self.skip_save_progress
+        skip_save_log = self.skip_save_log
+        # Set temporary values
+        self.skip_save_model = True
+        self.skip_save_progress = True
+        self.skip_save_log = True
+
+        # Turn eager mode on
+        tf.config.experimental_run_functions_eagerly(True)
```

Comment:
We should probably wrap this in a try-finally just in case an exception is raised and handled by the user. So something like:

```
tf.config.experimental_run_functions_eagerly(True)
try:
    ...
finally:
    tf.config.experimental_run_functions_eagerly(False)
```